test/alu_8bit_sub_test.c

```diff
@@ -0,0 +1,258 @@
+#include <criterion/criterion.h>
+
+#include "fixtures/cpu_mmu.h"
+#include "logging.h"
+
+Test(sub_a_n, sub_a_a_no_borrow, .init = cpu_mmu_setup, .fini = cpu_teardown) {
+    const uint8_t opcode = 0x97;
+    const uint8_t a = 128;
+    const uint8_t expected = 0;
+    const uint16_t address = rand() % (0xFFFF);
+
+    // setup cpu
+    cpu.PC = address;
+    cpu.AF.bytes.high = a;
+    mmu_write_byte(address, opcode);
+
+    // do the actual emulation
+    cpu_step();
+    const uint8_t actual = cpu.AF.bytes.high;
+
+    // Z should be set, as our result is 0
+    cr_assert(cpu.AF.bytes.low & (1 << Z_FLAG));
+
+    // N should be set
+    cr_assert(cpu.AF.bytes.low & (1 << N_FLAG));
+
+    // H should NOT be set because of no half-borrow
+    cr_assert(!(cpu.AF.bytes.low & (1 << H_FLAG)));
+
+    // C should NOT be set because of no borrow
+    cr_assert(!(cpu.AF.bytes.low & (1 << C_FLAG)));
+
+    // check if value is correct
+    cr_assert(actual == expected);
+
+    // check if pc is Pc is updated correctly
+    cr_assert(cpu.PC == (address + 1));
+}
+
+Test(sub_a_n, sub_a_b_only_half_borrow, .init = cpu_mmu_setup, .fini = cpu_teardown) {
+    const uint8_t opcode = 0x90;
+    const uint8_t a = 0b0100101;
+    const uint8_t b = 0b0001101;
+    const uint8_t expected = a - b;
+    const uint16_t address = rand() % (0xFFFF);
+
+    // setup cpu
+    cpu.PC = address;
+    cpu.AF.bytes.high = a;
+    cpu.BC.bytes.high = b;
+    mmu_write_byte(address, opcode);
```

Comment:
Very opinionated suggestion, so don't worry too much about it.

I think that this could be bundled in some static function to reduce the duplication among these tests:
```C
static uint8_t foo(const uint8_t opcode, const uint8_t lhs, const uint8_t rhs, const Register r) {
    const uint_fast16_t expected = lhs - rhs;
    const uint16_t address = rand() % (0xFFFF);

    // setup cpu
    cpu.PC = address;
    cpu.AF.bytes.high = lhs;
    r.bytes.high = rhs;
    mmu_write_byte(address, opcode);

    // do the actual emulation
    cpu_step();
    const uint8_t actual = cpu.AF.bytes.high;
    
    return (uint8_t) expected;
}
```
I hope I abstracted correctly, let me know what you think ^^

Reply:
The problem is that you cannot unify everything for the tests, e.g. sometimes we have `r.bytes.high?` but other times `r.bytes.low`.
Additionally, how would we do our assertions? Because it is bad style to outsource the assertions out of the test themselves
But we should probably keep this in mind if we have a lot of duplicated lines in our test code down the line.